### DISCUSSION TITLE
Multiple dockers running strongswan on the same host

### DISCUSSION THREAD

**Discussion by sharonenoch** [strongswan/strongswan#1066] [General]
We are trying to simulate testing of a one to many ipsec tunnel test with dockers and application traffic.. 

1 server (responder) and 24 clients (initiator) all using dockers, but with all dockers on the same host. Each docker is running an instance of strongswan.. Mostly it creates fine when initiated 1 at a time.. and tunnels establish from 25 clients to the server.. but at times the "ipsec status" does not respond for over 60 to 80 seconds across all the dockers.. This mainly happens when we initiate all client connections to the server at the same time or rekeying happening at the same time.. We see a lot half open IKE SAs and unnamed connections in the ipsec status output

We tried increasing the charon -> threads, hash table size, segments, stroke -> max concurrent jobs, stroke timeout.. but it still goes into that state.. Even after deleting all the tunnels, the ipsec status all showed all the tunnels.. it was like it was deadlocked or so.. below were the workers and medium priority jobs which kept increasing and reached 11269.. I had to restart strongswan service to come out of the state and then it cleared all the connections

  worker threads: 0 of 24 idle, 5/6/13/0 working, job queue: 0/3/11269/0, scheduled: 127

My question would be that, is this configuration theoretically correct? Since all the dockers being on the same host share the same kernel? or Is there any shared resource or structure or state mapping that the ipsec charon daemon is using with the kernel which causes this race condition in this particular config?

The reason for this doubt is initially we only thought the server was where it was happening since it had about 48 SAs (2 paths) getting established at the same time but then the clients which have only 2 ipsec SAs connections to the server also exhibit this issue..

**Comment by tobiasbrunner** [comment]
Did you try this with 5.9.6?

> Since all the dockers being on the same host share the same kernel? 

That's definitely a concern as SAD/SPD modifications will use shared locks. If you want to only test strongSwan and its config, you could use the [load-tester plugin](https://docs.strongswan.org/docs/5.9/plugins/load-tester.html)'s fake kernel interface.

**Reply by sharonenoch** [reply]
We are still running on Ubuntu 18.04, so using the distribution version 5.6.2.  Thank you for letting us know about the shared locks for the SAD and SPD.. 

Thanks, I was aware of the load-test plugin but since the intention was not to test Strongswan, so we had not used it.. We were trying to test all the functionality related to specific custom routing and application traffic routing via the ipsec tunnels 

We will use VMs now instead to test the same..
